### DISCUSSION TITLE
Next steps for DBT

### DISCUSSION THREAD

**Discussion by litui** [SynthstromAudible/DelugeFirmware#624] [General]
Now that DBT is in `community` there is a laundry list of stuff I want to add to it but I'll start with the following:

- [x] Disable default build in favour of listing targets.
- [ ] Build stats on completion, sizing, etc.
    - [ ] Danger-checking/warning for .bin size 0 or oversize .bin
    - [x] Output .siz file as e2studio does
- [x] Test for or just provide some notice that libcrypt.so.1 is needed, maybe with some instructions.
- [x] Dynamic version string (eg: "c20230619-a1b2c3d-d" in the scheme: &lt;c for community&gt;&lt;yyyyMMdd&gt;-&lt;git_commit&gt;-&lt;build type: "r" or "d"&gt;)
    - [x] DBT pre-runner for preparing things before a build, to be called by e2 studio and other editors prior to building.
- [x] Direct target support / multitarget (eg: `./dbt dbt-build-debug-oled dbt-build-release-oled` etc.)
    - [x] Fix compilation db (compile_commands.json) to output to the build directories
- [ ] Debugger shortcuts
    - [x] DelugeProbe/CMSIS-DAP and FTDI compatible openocd + gdb shortcuts
    - [ ] JLink support with whatever tools they use (know very little about JLink, but have some starting points from over at the Flipper Zero repo)
    - [ ] Quick, commandline "install" support via debugger-based hardware flashing
- [x] Restructure to make more modular in prep for all the above ^^^^^
- [x] Touch .map file so it gets cleaned up with the rest of the targets.
- [ ] clang-format runner
- [x] DBT-Toolchain v8 changes
    - [x] Include clang-format
    - [x] Remove cmake
    - [x] Fix mac dotfile pollution
    - [x] Increment version in bash/batch script downloader.
- [x] --nowo, --no_owo, --no_uwu (remove progress indicator) param

More to be added.

Playing around with Kconfiglib is on my to-do list here as well, but I'm not sure I have a specific goal in mind with it yet or idea of how it fits. Once I do, expect some checkboxes to be added.

Pie-in-the-sky / requires a lot of other things to be in place:

- Test and/or test framework runner
- UF2 image support (should work with DelugeProbe and maybe even USB Mass Storage eventually...)
- USB interface?
    - Maybe some midi sender/receiver mocking capability for putting the Deluge through its paces
    - DFU loader support

**Comment by m-m-adams** [comment]
Copying from the PR - on ARM Fedora 37 I get an error about missing libcrypt.so, but it builds successfully after installing libxcrypt-compat. This is arguably a distro issue but probably worth noting it as a requirement since this appears to affect Arch and WSL2 as well per this discussion in python-build-standalone: https://github.com/indygreg/python-build-standalone/issues/173

Otherwise I guess it will go away once everyone is running python 3.13


So confirmed working on Fedora 37 ARM!

**Comment by litui** [comment]
I'll add a bullet for that. I was thinking about the best way to approach testing for that and providing a notification if it's missing. Maybe something like:

 **if** the OS is Linux, grep `ldconfig -p` and determine if libcrypt.so.1 is in place/what we're expecting. Could even maybe ldd the file if we need to dig deeper.

eg `ldconfig -p | grep "libcrypt.so.1"` output
```
libcrypt.so.1 (libc6,x86-64) => /lib/x86_64-linux-gnu/libcrypt.so.1
```

**Comment by bfredl** [comment]
> This is arguably a distro issue but probably worth noting it as a requirement since this appears to affect Arch and WSL2 as well per this discussion in python-build-standalone: https://github.com/indygreg/python-build-standalone/issues/173

A possible solution could be to use the system's `python3.11` executable if it is available, and set up a venv to still install the packages into an isolated environment.

**Comment by litui** [comment]
Possible, not my preference though. Venvs are useful, but that's a whole other pile of logic and additional layers of path mangling magic I'd want to avoid for simplicity's sake. In one of the earlier iterations of DBT I tried bundling poetry wheels and using the system python, and it was messier than I'd like.

**Comment by litui** [comment]
Not done with this issue yet!
